python_modules/dagster/dagster/_core/definitions/asset_automation_evaluator.py

```diff
@@ -0,0 +1,227 @@
+        ...
+
+    def __and__(self, other: "AutomationCondition") -> "AutomationCondition":
+        # group AndAutomationConditions together
+        if isinstance(self, AndAutomationCondition):
+            return AndAutomationCondition(children=[*self.children, other])
+        return AndAutomationCondition(children=[self, other])
+
+    def __or__(self, other: "AutomationCondition") -> "AutomationCondition":
+        # group OrAutomationConditions together
+        if isinstance(self, OrAutomationCondition):
+            return OrAutomationCondition(children=[*self.children, other])
+        return OrAutomationCondition(children=[self, other])
+
+    def __invert__(self) -> "AutomationCondition":
+        return InvertAutomationCondition(child=self)
+
+
+class RuleCondition(
+    AutomationCondition, NamedTuple("_RuleCondition", [("rule", AutoMaterializeRule)])
+):
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        context.daemon_context._verbose_log_fn(f"Evaluating rule: {self.rule.to_snapshot()}")  # noqa
+        results = self.rule.evaluate_for_asset(context)
+        true_subset = context.empty_subset()
+        for _, asset_partitions in results:
+            true_subset |= AssetSubset.from_asset_partitions_set(
+                context.asset_key, context.partitions_def, asset_partitions
+            )
+        context.daemon_context._verbose_log_fn(f"Rule returned {true_subset.size} partitions")  # noqa
+        return ConditionEvaluation(condition=self, true_subset=true_subset, results=results)
+
+
+class AndAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_AndAutomationCondition", [("children", Sequence[AutomationCondition])]),
+):
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluations: List[ConditionEvaluation] = []
+        true_subset = context.candidate_subset
+        for child in self.children:
+            context = context.with_candidate_subset(true_subset)
+            result = child.evaluate(context)
+            child_evaluations.append(result)
+            true_subset &= result.true_subset
+        return ConditionEvaluation(
+            condition=self, true_subset=true_subset, children=child_evaluations
+        )
+
+
+class OrAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_OrAutomationCondition", [("children", Sequence[AutomationCondition])]),
+):
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluations: List[ConditionEvaluation] = []
+        true_subset = context.empty_subset()
+        for child in self.children:
+            result = child.evaluate(context)
+            child_evaluations.append(result)
+            true_subset |= result.true_subset
+        return ConditionEvaluation(
+            condition=self, true_subset=true_subset, children=child_evaluations
+        )
+
+
+class InvertAutomationCondition(
```

Comment:
While sketching out what later parts of this stack will look like, I realized that it's easiest to lean into the recursive structure of these expressions and say that all `AutomationConditions` will have a `children` property (which may of course be an empty list).

With that in mind, we can just make this a `NorAutomationCondition` (UI analog: "None of ..."), and allow it to support the negation of multiple things at once (for example, if you do `~(a | b)` we have the option to transform that into `Nor(a, b)` rather than `Nor(Or(a, b))`)